### PR TITLE
GLM-4.5 model accuracy alignment

### PR DESCRIPTION
**Pull request by zhengshengning**
### PR Category

Distributed Strategy

### PR Types

New features

### Description

GLM-4.5 模型精度对齐（PaddleFleet vs Megatron-LM），所有对齐逻辑通过 use_accuracy_compatible 开关控制，开关关闭时不影响原有代码行为。

核心改动：

moe_utils.py — 实现确定性 permute/unpermute（gather+sum fp32 PyLayer），消除 scatter_add 非确定性累加序差异

moe_router.py — gate weight dtype 改为 bf16（对齐 MG 默认行为）；router scoring 使用 re-sigmoid 计算 gates_ori；TopK 归一化使用 fp64 求和；aux_loss 计算对齐 MG 公式；take_along_axis 替换为 gather_nd（对齐反向梯度路径）

moe_layer.py — 三路 clone（ThreePathCloneAlignMG PyLayer）控制 bf16 backward 累加顺序；小 M expert padding（M<17 → 32）消除 cublasLt tiny-M kernel 差异；EP>1 时强制 alltoall dispatcher

fused_bias_swiglu.py — 对齐模式下使用 eager SwiGLU forward/backward（非 fused），避免 jit_fuser kernel 数值差异

tensor_parallel/layers.py — forward 使用 weight.T.contiguous() + paddle.matmul 对齐 GEMM layout；expert weight 的 wgrad 使用 fp32 直接累加到 main_grad

language_loss.py — loss 求和使用 fp64 精度；模拟 EP allreduce 的 bf16 累加序

model_parallel_config.py — 对齐模式下自动启用 deterministic_mode=True

rotary_pos_embedding.py — 移除已无用的 use_accuracy_compatible 分支（CPU/GPU 计算 inv_freq 数值等价）

### 是否引起精度变化

否

### CODE REVIEW

File: src/paddlefleet/transformer/moe/moe_layer.py [hunk 1]
```diff
@@ -1071,6 +1106,23 @@ def fusion_moe_forward(
         topk_weights: paddle.Tensor | None = None,
         topk_indices: paddle.Tensor | None = None,
     ):
+        if self.use_accuracy_compatible:
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/moe/moe_layer.py:1221]: P1 `use_accuracy_compatible=True` 且 `using_sonic_moe=True` 时，这里直接转入 `custom_forward`，但 SonicMoE 初始化只创建 `grouped_gemm_experts`，`_use_grouped_mlp_expert=False`，不会创建 `self.experts`。随后 `custom_forward -> routed_experts_compute -> expert_forward` 会访问不存在的 `self.experts`，并在首个 MoE 前向抛出 `AttributeError`。请保留 SonicMoE 的 grouped/fused 执行路径，或在配置初始化阶段明确拒绝该组合。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
### Paddle-Bot Review Board （review完成)

| 序号 | 位置 | 优先级 | 规则来源 | 状态 |
| --- | --- | --- | --- | --- |
| 1 | [SonicMoE 准确率模式前向](https://github.com/PaddlePaddle/PaddleFleet/pull/1903#discussion_r3880492010) | P1 | [仓库规则：功能正确性与兼容性](https://github.com/PaddlePaddle/PaddleFleet/blob/9f9bd7b9952d3cbfdd48f80138cf4d794c753979/.agents/skills/ai-review/references/base-rules.md) | ✅ |
| 2 | [GLM45Air 对齐数据顺序](https://github.com/PaddlePaddle/PaddleFleet/pull/1903#discussion_r3880670448) | P1 | [仓库规则：功能正确性与兼容性](https://github.com/PaddlePaddle/PaddleFleet/blob/9f9bd7b9952d3cbfdd48f80138cf4d794c753979/.agents/skills/ai-review/references/base-rules.md) | ✅ |
| 3 | [setup_venvs.sh 参数展开](https://github.com/PaddlePaddle/PaddleFleet/pull/1903#discussion_r3894836347) | P0 | [仓库规则：构建与依赖正确性](https://github.com/PaddlePaddle/PaddleFleet/blob/9f9bd7b9952d3cbfdd48f80138cf4d794c753979/.agents/skills/ai-review/references/base-rules.md) | 🚧 |
| 4 | [对齐脚本默认 wheel 地址](https://github.com/PaddlePaddle/PaddleFleet/pull/1903#discussion_r3894838426) | P1 | [仓库规则：构建与依赖正确性](https://github.com/PaddlePaddle/PaddleFleet/blob/9f9bd7b9952d3cbfdd48f80138cf4d794c753979/.agents/skills/ai-review/references/base-rules.md) | 🚧 |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.6-sol 默认推理级别, please check the suggestions carefully.</sup>
</div>

File: scripts/alignment_model_accuracy/GLM45Air_EP2/GLM45Air_EP2.yaml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+eval_dataset_path: /home/.cache/PaddleFormers/GLM-4.5-Air-tiny-2L/alignment_paddle.jsonl
+train_dataset_prob: "1.0"
+eval_dataset_prob: "1.0"
+
+# 数据处理
+packing: false
+truncate_packing: false
+max_seq_len: 8192
+mix_strategy: concat
+random_shuffle: true
```
Comment (risemeup1111) [inline scripts/alignment_model_accuracy/GLM45Air_EP2/GLM45Air_EP2.yaml]: P1 这个新增对齐用例在 Paddle 侧启用了数据集随机打乱，而对应 Torch 脚本同时设置了 `--dataset_shuffle False` 和 `--train_dataloader_shuffle False`。`compare_loss.py` 又按日志出现顺序逐 step 比较 MD5，因此只要数据集含多个样本，两端就可能在同一步读取不同样本，结果无法区分数据顺序差异和实现精度差异。请将 Paddle 侧设为 `false`，或让两侧采用同一套可验证的采样序列后再纳入 `CASES`。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

**Comment by risemeup1111** [comment]
## Failed CI看板

| 流水线名称 | 问题标签 | 修复建议 | 日志片段 |
| ------------ | ---------- | ---------- | ---------- |
| Check approval | 路径审批门禁未满足 | 为 `src/paddlefleet/model_parallel_config.py` 获取要求的审批后重新运行审批检查。 | [报错代码](https://github.com/PaddlePaddle/PaddleFleet/actions/runs/33171838866/job/98850625054) |
| Require review-bot approval | 当前 head 缺少机器人或指定审批人批准 | 针对当前 SHA 完成 review-bot approve 或取得指定审批人批准。 | [报错代码](https://github.com/PaddlePaddle/PaddleFleet/actions/runs/33171838909/job/98850736699) |
| Unit test (single card) | 单卡测试超时 | 定位超时前最后运行的测试并单独执行新增测试及其相关测试集，修复挂起或性能回退后重跑。 | [报错代码](https://github.com/PaddlePaddle/PaddleFleet/actions/runs/33171838987/job/99046305786) |
| model accuracy alignment (H20, multi-card) | Python 依赖解析失败 | 检查 `setup_venvs.sh` 中 `mcore-bridge`、Paddle 索引和 `paddlefleet-ops` wheel 的版本约束并锁定可解析组合。 | [报错代码](https://github.com/PaddlePaddle/PaddleFleet/actions/runs/33171838987/job/99046305801) |
| Coverage Upload And Check | 上游覆盖率文件缺失，具体检查失败行未包含在快照中 | 先让单卡和 H20 任务生成预期 XML，再重跑覆盖率检查，并将未参与矩阵的报告从必需输入中移除或显式跳过。 | [报错代码](https://github.com/PaddlePaddle/PaddleFleet/actions/runs/33171838987/job/99061405599) |

<details>
<summary>日志分析报告</summary>

**失败的测试 case**:
```text
1. Check approval
   步骤：检查变更文件的路径审批规则。
   关键错误：model_parallel_config.py 需要 sneaxiy、From00 之一审批；存在 1 个未满足的审批错误；exit code 6。

2. Require review-bot approval
   步骤：读取当前 head dc034fa1d0690b7a9fb493f88cf97e0016ea9cac 上的评审状态。
   关键错误：review-bot risemeup1111 尚未 APPROVED，指定审批账号也未满足门禁；exit code 1。

3. Unit test (single card)
   步骤：启动单卡测试容器并执行测试套件。
   关键错误：Single card test 在 120 分钟后超时；快照没有显示断言失败或明确挂起的测试名。

4. model accuracy alignment (H20, multi-card)
   步骤：进入 alignment_model_accuracy，配置环境并解析依赖。
   关键错误：依赖解析报告 requirements are unsatisfiable，随后 wrapper 输出 Model accuracy align failed；模型对齐 case 尚未实际执行。

5. Coverage Upload And Check
   步骤：收集并上传 multi、single、H20 等 coverage XML。
   关键错误：single_coverage.xml、h20_multi_coverage.xml 等文件不存在；Codecov upload 本身显示 success，快照未提供最终 Coverage Check 的具体失败行。
```

**根本原因分析**

两个审批 job 属于同一类 PR 治理门禁，但相互独立：PR 修改了受保护的 `model_parallel_config.py`，路径审批未满足；同时当前 head 没有 review-bot 的决定性 `APPROVED` 评审。这些不是代码运行时失败。

H20 多卡任务的可确认根因是环境依赖不可解析。镜像拉取成功，失败发生在统一环境配置阶段，尚未运行 `MinimaxV2.5_EP2` 或新增的 `GLM45Air_EP2` 对齐脚本。PR 同时修改了 `setup_venvs.sh`：将 `mcore-bridge` 从版本号 `1.4.3` 改为 `1.7.0.dev0` wheel URL，并改变了 Paddle 包索引和 `--no-config` 参数。这些改动是最直接的回归范围，但由于冲突详情被裁剪，不能进一步确定具体冲突的包对。

单卡任务的直接原因是超过 120 分钟动作超时。日志中 `test_ai_arguments.py` 的 `--unknown-flag` 输出是测试预期的参数解析错误，不能作为失败根因；快照也没有提供最后一个未完成测试，因此不能据此断定是某个新增测试或某段模型代码造成挂起。PR 新增的 accuracy-compatible 分支及测试应作为优先排查范围。

Coverage job 更像是上游失败的次生结果：单卡超时可解释 `single_coverage.xml` 缺失，H20 依赖解析失败可解释 `h20_multi_coverage.xml` 缺失。快照只显示 Codecov 上传成功和缺失文件警告，不能证明 Codecov 网络或上传失败，也不足以确定最终覆盖率阈值检查的具体原因。

**修复建议**

1. 先修复 `setup_venvs.sh` 的依赖解析问题：逐项验证 `mcore-bridge`、`paddlefleet-ops`、Paddle wheel 及索引配置，锁定兼容版本；重点对比本次新增的 `mcore-bridge` URL、`--no-config` 和显式 Paddle index 改动。
2. 对单卡任务使用详细测试输出和耗时统计定位超时点，优先单独运行本次新增或修改的测试文件，再检查 `use_accuracy_compatible` 的初始化、MoE、loss 和 GEMM 分支是否引入阻塞。
3. 获得路径审批及当前 head 的机器人审批后，重新运行两个审批 job。
4. 依赖问题解决后重跑 `model accuracy alignment (H20, multi-card)`；单卡超时修复后重跑 `Unit test (single card)`。
5. 待上游任务生成覆盖率文件后，再重跑 `Coverage Upload And Check`；若部分矩阵本来就未启用，应调整覆盖率检查的必需文件列表。

</details>

Powered by Nyanpasu with gpt-5.6-luna 默认推理级别, please check the suggestions carefully.

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已确认当前提交修复了 GLM45Air 两侧数据乱序问题；SonicMoE 与准确率兼容模式的组合仍会进入依赖 `self.experts` 的非融合路径，原 P1 尚未解决，因此本轮仍请求修改。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.6-sol 默认推理级别, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
当前 head 仅更新了对齐环境安装脚本，未改变 SonicMoE 路径；`use_accuracy_compatible=True` 与 `using_sonic_moe=True` 时仍会进入依赖不存在的 `self.experts` 的路径，原 P1 仍未解决，因此继续请求修改。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.6-sol 默认推理级别, please check the suggestions carefully.</sup>
</div>

File: scripts/alignment_model_accuracy/setup_venvs.sh [hunk 3]
```diff
@@ -33,6 +32,7 @@ readonly PADDLEFLEET_OPS_WHEEL="${PADDLEFLEET_OPS_WHEEL_PATH:-${NIGHTLY_WHL_BASE
 readonly PADDLEFORMERS_WHEEL="${NIGHTLY_WHL_BASE}/paddleformers/paddleformers-0.0.0.dev-py3-none-any.whl"
 readonly MEGATRON_CORE_WHEEL="${MEGATRON_CORE_WHEEL_PATH:-${NIGHTLY_WHL_BASE}/megatron_core-0.19.0+f2706b6f3-cp312-cp312-linux_x86_64.whl}"
 readonly MS_SWIFT_WHEEL="${MS_SWIFT_WHEEL_PATH:-${NIGHTLY_WHL_BASE}/ms_swift-4.5.0.dev0-py3-none-any.whl}"
+readonly MCORE_BRIDGE_WHEEL="${MCORE_BRIDGE_WHEEL_PATH:-${NIGHTLY_WHL_BASE}/mcore_bridge-1.7.0.dev0-py3-none-any.whl"
```
Comment (risemeup1111) [inline scripts/alignment_model_accuracy/setup_venvs.sh]: P0 `setup_venvs.sh` 第 35 行的参数展开缺少外层 `}`，脚本在解析阶段即报 `unexpected EOF while looking for matching "`，因此 `run_alignment_test.sh` 无法创建任何虚拟环境。请补上 `${...}` 的闭合括号，并运行 `bash -n scripts/alignment_model_accuracy/setup_venvs.sh` 验证。

File: scripts/alignment_model_accuracy/setup_venvs.sh [hunk 4]
```diff
@@ -33,6 +32,7 @@ readonly PADDLEFLEET_OPS_WHEEL="${PADDLEFLEET_OPS_WHEEL_PATH:-${NIGHTLY_WHL_BASE
 readonly PADDLEFORMERS_WHEEL="${NIGHTLY_WHL_BASE}/paddleformers/paddleformers-0.0.0.dev-py3-none-any.whl"
 readonly MEGATRON_CORE_WHEEL="${MEGATRON_CORE_WHEEL_PATH:-${NIGHTLY_WHL_BASE}/megatron_core-0.19.0+f2706b6f3-cp312-cp312-linux_x86_64.whl}"
```
Comment (risemeup1111) [inline scripts/alignment_model_accuracy/setup_venvs.sh:34]: P1 这里将 Megatron/ms-swift/mcore-bridge 的默认 wheel 改到 `paddle-whl.bj.bcebos.com/nightly/cu130/` 根路径，但当前这三个精确 URL 均返回 404；CI 虽通过环境变量覆盖，README 中不设置覆盖变量的本地默认流程仍会在安装阶段失败。请改回可用的 wheel 地址，或提供并验证 nightly 中实际存在的路径。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
新增对齐环境脚本仍有两个未解决问题：`setup_venvs.sh` 第 35 行缺少参数展开的闭合 `}`，会导致脚本解析失败；同时 Megatron/ms-swift/mcore-bridge 的默认 wheel 地址返回 404。请修复脚本语法并改用可验证存在的默认 wheel 地址后再合入。此前的 SonicMoE 组合问题和 GLM45Air 数据顺序问题已在当前目标提交中分别通过显式拒绝和关闭随机打乱处理。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.6-sol 默认推理级别, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复核当前提交，以下问题仍未解决：

- `scripts/alignment_model_accuracy/setup_venvs.sh` 第 35 行的 `${...}` 外层 `}` 仍缺失；`bash -n` 报 `unexpected EOF`，脚本无法执行。
- 同脚本中 Megatron、ms-swift、mcore-bridge 的默认 wheel 地址仍返回 404；未设置覆盖变量的 README 默认流程会在安装阶段失败。

请修复脚本语法并改用可验证存在的默认 wheel 地址后再合入。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.6-sol 默认推理级别, please check the suggestions carefully.</sup>
</div>

File: scripts/alignment_model_accuracy/setup_venvs.sh (hunk 4 above)
Comment (zhengshengning) [reply scripts/alignment_model_accuracy/setup_venvs.sh:34]: 本PR这个地方啥也没改，请认真review!

**Comment by zhengshengning** [comment]
@Paddle-Bot 请review代码，完成approve

**Review by From00 — APPROVED** [review]
LGTM
